[0.34] DAG only deploys (#3269)
astronomer/docs: software/deploy-cli.md

---

> sidebar_label: 'Deploy code using the CLI'

Author: ```suggestion
sidebar_label: 'Deploy a project image'
```

title: 'Deploy code to Astronomer Software using the Astro CLI'
id: deploy-cli
description: How to push DAGs to your Airflow Deployment on Astronomer Software using the Astro CLI.
---

> To run your code on Astronomer Software, you need to deploy it to a Deployment. You can deploy part or all of an Astro project to an Astro Deployment using the Astro CLI.

Reviewer: Above I see us using both "Airflow Deployment" and the bare "Deployment", and I feel like being more specific makes it easier to understand. Also, shouldn't we avoid using "Astro" since this is not at all related to that product?
```suggestion
To run your code on Astronomer Software, you need to deploy it to an Airflow Deployment. You can deploy part or all of an Astronomer project to an Airflow Deployment using the Astro CLI.
```

> When you deploy a project, the Astro CLI builds your all files in your Astro project, including DAGs, into a Docker image. It then pushes this mage to an image registry on your Astronomer Software cluster where your Deployment can access the image and apply it to its running Airflow containers.

Reviewer: Again we're using "Astro" here. Our product naming sure makes it hard to talk about our products in a clear way. :(

> When you deploy a project, the Astro CLI builds your all files in your Astro project, including DAGs, into a Docker image. It then pushes this mage to an image registry on your Astronomer Software cluster where your Deployment can access the image and apply it to its running Airflow containers.

Reviewer: ```suggestion
When you deploy a project, the Astro CLI builds your all files in your Astro project, including DAGs, into a Docker image. It then pushes this image to an image registry on your Astronomer Software cluster where your Deployment accesses the image and uses it to run Airflow containers.
```

For guidance on automating this process, refer to [Deploy to Astronomer via CI/CD](ci-cd.md). To learn how to add Python and OS-level packages or otherwise customize your Docker image, read [Customize your image](customize-image.md).

This document covers deploying a complete project image to Astro. To deploy DAGs only, see:

- [Deploy DAGs using the Astro CLI](deploy-dags.md)
- [Deploy DAGs to an NFS volume](deploy-nfs.md)

> - [Deploy DAGs using git-sync](deploy-nfs.md).

Reviewer: No line item for dag-deploy?

astro deploy
```

> This command returns a list of Airflow Deployments available in your Workspace and prompts you to pick one. After you execute the command, all files in your Astro project directory are built into a new Docker image and Docker containers for all Airflow components are restarted.

Reviewer: ```suggestion
This command returns a list of Airflow Deployments available in your Workspace and prompts you to pick one. After you execute the command, all files in your Astronomer project directory are built into a new Docker image, the image is pushed to the Astronomer Software registry, and the Docker containers for all Airflow components in the Airflow Deployment are restarted.
```

> This command returns a list of Airflow Deployments available in your Workspace and prompts you to pick one. After you execute the command, all files in your Astro project directory are built into a new Docker image, the image is pushed to the Astronomer Software registry, and the Docker containers for all Airflow components in the Deployment are restarted.

Reviewer: IMHO we should reduce the usage of the word "Docker" overall, and especially when we are talking about k8s, because Docker is not part of k8s. Technically these are OCI containers, and they run in containerd. k8s used to use docker, but no longer does. OCI containers were defined using the Docker container spec, so they are related to Docker, but modern containers are technically not "docker containers"
```suggestion
This command returns a list of Airflow Deployments available in your Workspace and prompts you to pick one. After you execute the command, all files in your Astro project directory are built into a new Docker image, the image is pushed to the Astronomer Software registry, and the containers for all Airflow components in the Deployment are restarted.
```

For more information about what gets built into your image, read [Customize your image](customize-image.md).

## Deploy DAGs only using the Astro CLI

DAG-only deploys are the fastest way to deploy code to Astronomer Software. They are recommended if you only need to deploy changes made to the `dags` directory of your Astro project.

> DAG-only deploys are enabled by default on all Deployments on Astronomer Software. When they are enabled, you must still do a full project deploy when you make a change to any file in your Astro project that is not in the `dags` directory, or when you [upgrade Astro Runtime](manage-airflow-versions.md).

Author: ```suggestion
When this feature is enabled, you must still do a full project deploy when you make a change to any file in your Astro project that is not in the `dags` directory, or when you [upgrade Astro Runtime](manage-airflow-versions.md).
```

DAG-only deploys have the following benefits:

- DAG-only deploys are significantly faster than project deploys.
- Deployments pick up DAG-only deploys without restarting. This results in a more efficient use of workers and no downtime for your Deployments.
- If you have a CI/CD process that includes both DAG and image-based deploys, you can use your repository's permissions to control which users can perform which kinds of deploys.
- You can use DAG deploys to update your DAGs when you have slow upload speed on your internet connection.

### Enable the feature

> [Insert instructions]

Author: Need tech review here, not sure what the config is for this

Reviewer: this is going to be a GUI configured feature that is enabled or disabled per-airflow-deployment. @pgvishnuram is the SME for the astronomer chart features.

Reviewer 2: customer can leverage both api and ui for enabling or disabling  dag_deploy for per-airflow-deployment
